internal/pkg/agent/application/upgrade/step_download_test.go

```diff
@@ -0,0 +1,147 @@
+
+		logs := obs.TakeAll()
+		require.Len(t, logs, 1)
+		require.Equal(t, fmt.Sprintf("download attempt 1 of %d", settings.RetryMaxCount+1), logs[0].Message)
+	})
+
+	// Downloader constructor failing on first attempt, but succeeding on second attempt (= first retry)
+	t.Run("constructor_failure_once", func(t *testing.T) {
+		attemptIdx := 0
+		mockDownloaderCtor := func(version string, log *logger.Logger, settings *artifact.Config) (download.Downloader, error) {
+			defer func() {
+				attemptIdx++
+			}()
+
+			switch attemptIdx {
+			case 0:
+				// First attempt: fail
+				return nil, errors.New("failed to construct downloader")
+			case 1:
+				// Second attempt: succeed
+				return &mockDownloader{expectedDownloadPath, nil}, nil
+			default:
+				require.Fail(t, "should have succeeded after 2 attempts")
+			}
+
+			return nil, nil
+		}
+
+		u := NewUpgrader(testLogger, &settings, &info.AgentInfo{})
+		path, err := u.downloadWithRetries(context.Background(), mockDownloaderCtor, "8.9.0", &settings)
+		require.NoError(t, err)
+		require.Equal(t, expectedDownloadPath, path)
+
+		logs := obs.TakeAll()
+		require.Len(t, logs, 3)
+		require.Equal(t, fmt.Sprintf("download attempt 1 of %d", settings.RetryMaxCount+1), logs[0].Message)
+		require.Contains(t, logs[1].Message, fmt.Sprintf("unable to create fetcher: failed to construct downloader; retrying (will be retry 1 of %d)", settings.RetryMaxCount))
+		require.Equal(t, fmt.Sprintf("download attempt 2 of %d", settings.RetryMaxCount+1), logs[2].Message)
+	})
+
+	// Download failing on first attempt, but succeeding on second attempt (= first retry)
+	t.Run("download_failure_once", func(t *testing.T) {
+		attemptIdx := 0
+		mockDownloaderCtor := func(version string, log *logger.Logger, settings *artifact.Config) (download.Downloader, error) {
+			defer func() {
+				attemptIdx++
+			}()
+
+			switch attemptIdx {
+			case 0:
+				// First attempt: fail
+				return &mockDownloader{"", errors.New("download failed")}, nil
+			case 1:
+				// Second attempt: succeed
+				return &mockDownloader{expectedDownloadPath, nil}, nil
+			default:
+				require.Fail(t, "should have succeeded after 2 attempts")
+			}
+
+			return nil, nil
+		}
+
+		u := NewUpgrader(testLogger, &settings, &info.AgentInfo{})
+		path, err := u.downloadWithRetries(context.Background(), mockDownloaderCtor, "8.9.0", &settings)
+		require.NoError(t, err)
+		require.Equal(t, expectedDownloadPath, path)
+
+		logs := obs.TakeAll()
+		require.Len(t, logs, 3)
+		require.Equal(t, fmt.Sprintf("download attempt 1 of %d", settings.RetryMaxCount+1), logs[0].Message)
+		require.Contains(t, logs[1].Message, fmt.Sprintf("unable to download package: download failed; retrying (will be retry 1 of %d)", settings.RetryMaxCount))
+		require.Equal(t, fmt.Sprintf("download attempt 2 of %d", settings.RetryMaxCount+1), logs[2].Message)
+	})
+
+	// Download unsuccessful (all retries exhausted)
+	t.Run("download_unsuccessful", func(t *testing.T) {
```

Comment:
It looks like we are missing a case to cover the upgrade failing because the download timeout expired.

Ideally the test would verify multiple retries can occur within the download timeout.